controller/setting_controller.go

```diff
@@ -390,6 +395,85 @@ func (sc *SettingController) syncBackupTarget() (err error) {
 	return nil
 }
 
+func (sc *SettingController) handleSecretsForAWSIAMRoleAnnotation(backupTargetURL, oldSecretName, newSecretName string, isBackupTargetURLChanged bool) (err error) {
+	isSameSecretName := oldSecretName == newSecretName
+	if isSameSecretName && !isBackupTargetURLChanged {
+		return nil
+	}
+
+	isHavingAWSIAMRoleArn := false
+	if !isSameSecretName {
+		isHavingAWSIAMRoleArn, _, err = sc.updateSecretForAWSIAMRoleAnnotation(backupTargetURL, oldSecretName, true)
+		if err != nil {
+			return err
+		}
+	}
+	_, isNewSecretNotExist, err := sc.updateSecretForAWSIAMRoleAnnotation(backupTargetURL, newSecretName, false)
+	if err != nil {
+		return err
+	}
+	// kubernetes_secret_controller will not reconcile the secret that does not exist such as named "", so we remove AWS IAM role annotation of pods if new secret name is "".
+	if isNewSecretNotExist && isHavingAWSIAMRoleArn {
+		if err = sc.removePodsAWSIAMRoleAnnotation(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (sc *SettingController) updateSecretForAWSIAMRoleAnnotation(backupTargetURL, secretName string, isOldSecret bool) (bool, bool, error) {
+	if secretName == "" {
+		return false, true, nil
+	}
+	isHavingAWSIAMRoleArn := false
+
+	secret, err := sc.ds.GetSecret(sc.namespace, secretName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, true, nil
+		}
+		return false, false, err
+	}
+
+	if isOldSecret {
+		delete(secret.Annotations, types.GetLonghornLabelKey(string(types.SettingNameBackupTarget)))
+		isHavingAWSIAMRoleArn = secret.Data[types.AWSIAMRoleArn] != nil
+	} else {
+		secret.Annotations[types.GetLonghornLabelKey(string(types.SettingNameBackupTarget))] = backupTargetURL
+	}
+
+	if _, err = sc.ds.UpdateSecret(sc.namespace, secret); err != nil {
+		return false, false, err
+	}
+	return isHavingAWSIAMRoleArn, false, nil
+}
+
+func (sc *SettingController) removePodsAWSIAMRoleAnnotation() error {
+	managerPods, err := sc.ds.ListManagerPods()
+	if err != nil {
+		return err
+	}
+
+	instanceManagerPods, err := sc.ds.ListInstanceManagerPods()
+	if err != nil {
+		return err
+	}
+	pods := append(managerPods, instanceManagerPods...)
+
+	for _, pod := range pods {
+		_, exist := pod.Annotations[types.AWSIAMRoleAnnotation]
+		if !exist {
+			continue
+		}
+		delete(pod.Annotations, types.AWSIAMRoleAnnotation)
+		sc.logger.Infof("Removing AWS IAM role for pod %v/%v", pod.Namespace, pod.Name)
```

Comment:
nit: do we need to log the pod.Namespace?

Reply:
Removed to log the pod.Namespace.